### ISSUE TITLE
Darling Is Unable To Run i386 Executables In 11.5 Source Update

### ISSUE THREAD

**Issue by CuriousTommy** [darlinghq/darling#1335]
**Expected Result**
The i386 application should print "Hello world!"

**Actual Result**
What did happen?
```bash
$ ./a.out 
*** 61:61: dserver_rpc_explicit_checkin: BAD SEND STATUS: -22 ***
Failed to checkin with darlingserver
```

**Steps To Reproduce**
1. Create the following hello world program
```c
#include <stdio.h>

int main() {
    printf("Hello world!\n");
    return 0;
}
```
2. Build a i386 executable in Darling (or on version of macOS that supports building 32bit apps)
```bash
clang -m32 main.c
```
3. Attempt to execute the i386 executable.

**System Information**
What system are you using?

| Software | Version |
| --- | --- |
| Linux Kernel | 6.1.14-200.fc37.x86_64 |
| Darling | 15e5b8024277d19a5a654256ab3ac7f767320aa4 `update_sources_11.5_alpharelease2` |

**Comment by CuriousTommy** [comment]
# Notes

**Comment by facekapow** [comment]
@CuriousTommy I just tested this on master and it turns out I get the exact same error:

```sh
$ darling --version
darling master @ b9f13c040
Copyright (C) 2012-2020 Lubos Dolezel
$ darling shell zsh
darling$ clang -o hello32 -arch i386 hello.c
ld: warning: The i386 architecture is deprecated for macOS (remove from the Xcode build setting: ARCHS)
darling$ clang -o hello64 -arch x86_64 hello.c
darling$ ./hello64
Hello, world!
darling$ ./hello32
*** 295:295: dserver_rpc_explicit_checkin: BAD SEND STATUS: -22 ***
Failed to checkin with darlingserver
```

**Comment by facekapow** [comment]
Given that it's not exclusive to the `update-sources` branch, it's most likely some structure size assumption I made when developing the RPC that holds for 64-bit platforms but breaks on 32-bit platforms.

**Comment by facekapow** [comment]
This should be fixed with the latest commit to master.

**Comment by facekapow** [comment]
I actually also ran into this on master while trying to get xtrace working on i386. This check is just for debugging that I forgot to remove before committing, but I think it's catching an actual issue: I believe the architecture is being reported incorrectly and the duct-tape code thinks it's an x86_64 process, so it tries to load the x86_64 state (which is wrong since this is an i386 process). I'll check that out tomorrow and debug it.

**Comment by facekapow** [comment]
Okay, the `bad RIP` panic should be fixed on master now. However, it does seem that i386 executables are broken on `update_sources_11.5` because the `bad RIP` panic was only occurring due to signal processing. So the process was likely running into a SIGSEGV or SIGABRT, doing signal processing with darlingserver, and then running into the `bad RIP` panic from having an incorrect architecture reported (which should now be fixed).

**Comment by CuriousTommy** [comment]
@facekapow Are you able to create a coredump with the `darling-coredump` util? I want to debug the `Segmentation fault: 11 (core dumped)`, but I'm unable to create a valid coredump:

```
$ darling-coredump mldr32_core_dump 
Input file is not a valid corefile
```

Trying to load the corefile doesn't work.

```
$ lldb --core darlingcore-mldr32_core_dump 
(lldb) target create --core "darlingcore-mldr32_core_dump"
error: Unable to find process plug-in for core file '/home/thomasa/Documents/CodingProjects/Repository/GitHub/darling/darlingcore-mldr32_core_dump'
```

**Comment by facekapow** [comment]
No, I haven't added support for 32-bit core dumps to `darling-coredump` yet.

**Comment by facekapow** [comment]
I was actually thinking of ditching darling-coredump in favor of an alternative approach where darlingserver can create Mach-O core dumps directly from a process while it's dying. The idea would be for the process to notify darlingserver it needs a core dump, stop itself (SIGSTP) before it dies, and then have darlingserver trace it to collect the necessary information for a core dump.

**Comment by CuriousTommy** [comment]
> I was actually thinking of ditching darling-coredump in favor of an alternative approach where darlingserver can create Mach-O core dumps directly from a process while it's dying. The idea would be for the process to notify darlingserver it needs a core dump, stop itself (SIGSTP) before it dies, and then have darlingserver trace it to collect the necessary information for a core dump.
> 
> The biggest issue with this approach is: what do you do if the signal processing code (sigproc in libsystem_kernel) has an error?

My general main concern is how long would it take to resolve/workaround the issues and implement this alternative approach? Because if it is going to take a while to get this resolved, I rather focus on adding i386 (and ARM64) support into `darling-coredump`.

**Comment by facekapow** [comment]
Yeah, that's also another issue; it's of course simpler to extend `darling-coredump` to support i386 (and ARM64 in the future). I could try to extend it with i386 support this week.

I'm just thinking in general about the core dump situation, because `darling-coredump` isn't very reliable at producing usable core dumps.

**Comment by CuriousTommy** [comment]
> Being able to run older applications is nice but doesn't this need a lot of workarounds?? (just asking)

Maybe... it's hard to tell how much work it would take to get i386 apps working again. Once I'm able to debug the i386 core dumps, I can figure out how to get this working.

If it does end up being too complicated to support i386, I might consider dropping i386 support (but I rather not do that...)

**Comment by CuriousTommy** [comment]
> No no no, let's please not do that.

To be clear, I would never do anything that drastic without consulting to the team first.

> We should support more stuff than Apple does, not less, including historical architectures like PPC. That's also what we've agreed on in the past.

If I ever did consider "dropping i386 support", the most I would do is temporarily disable i386 building until we figure out a solution to getting i386 executables to run. But like I said, I would always talk to the team first before doing anything that drastic.

> I obviously can't force you to do the work required for supporting i386; so if you decide you can't support it, I will have to invest time into it. But let's please not drop i386.

Don't worry, I'm still interesting in getting i386 working (although it may take a while to get it working since I'm busy with some IRL stuff).

**Comment by bugaevc** [comment]
Ah, awesome 👍
Yes, I can very much understand being busy / not having time (neither do I have time). You're doing great work, don't let me discourage you.

**Comment by facekapow** [comment]
With the latest commit on master, `darling-coredump` should work for 32-bit core dumps now.

**Comment by CuriousTommy** [comment]
> With the latest commit on master, `darling-coredump` should work for 32-bit core dumps now.

So I did a quick test on the `master` branch and was able to get a working core dump (I created a simple application that runs `abort`)!

Unfortunately, I was not able to able to analyze the core dump that was generated in `update_sources_11.5`. 

```
$ lldb --core darlingcore-core_dump 
(lldb) target create --core "darlingcore-core_dump"
Core file '/home/thomasa/Documents/CodingProjects/Repository/GitHub/darling/darlingcore-core_dump' (i386) was loaded.
(lldb) bt
* thread #1
  * frame #0: 0x000000e0
(lldb) frame select 0
frame #0: 0x000000e0
error: core file does not contain 0xe0
```
